{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# First - read the manuscript\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2951089/\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# What Can BioRuby Do?\n",
    "\n",
    "\n",
    "* Sequence and feature (e.g. gene, exon, intron) creating/reading/writing\n",
    "* Sequence database retrieval (e.g. embl, genbank, swissprot…)\n",
    "* Sequence Alignment (many many types: fasta, clustal, phylip…,..,..,..,..,...,...)\n",
    "* Sequence and feature database creation/query\n",
    "* Sequence search/analysis (e.g. blast, ‘virtual restrictions’, & many more)\n",
    "* Sequence annotation\n",
    "* Sequence assembly\n",
    "* Physical mapping (e.g. clones, contigs, etc.)\n",
    "* Taxonomy representation and database query/retrieval\n",
    "* Bio-ontology manipulation/exploration (e.g. Gene Ontology)\n",
    "* Phenotype representation and query (e.g. OMIM)\n",
    "* Phylogeny analyses\n",
    "* more more more more more more\n",
    "\n",
    "\n",
    "# We will start from the command-line\n",
    "\n",
    "BioRuby provides its own command line (i.e. its own \"terminal window\")\n",
    "\n",
    "    /Lectures/$  bioruby              < NOTE WHERE WE START!  IN THE LECTURES FOLDER!!!!\n",
    "    Creating directory (/home/osboxes/.bioruby/shell/session) ... done\n",
    "    Creating directory (/home/osboxes/.bioruby/shell/plugin) ... done\n",
    "    Creating directory (/home/osboxes/.bioruby/data) ... done\n",
    "\n",
    "    . . . B i o R u b y   i n   t h e   s h e l l . . .\n",
    "\n",
    "      Version : BioRuby 1.5.1 / Ruby 2.4.2\n",
    "\n",
    "    bioruby> \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demo of FASTQ Sequence quality masking\n",
    "\n",
    "It is important to mask low-quality base reads before doing a wide range of downstream operations like alignment or SNP-calling.\n",
    "\n",
    "We will now take a sequence file (in <code> Lectures/files/SP1.fq </code>) and we will use the BioRuby command line to mask it.\n",
    "\n",
    "The code to do this is:\n",
    "\n",
    "---------------------------\n",
    "    Bio::FlatFile.open('./files/SP1.fq').each do |entry|  # each sequence will be put into 'entry'\n",
    "        hq = entry.mask(20)                               # set the quality filter to 20\n",
    "        puts hq.output_fasta(entry.entry_id)              # print the FASTA out again, with the same ID as the head\n",
    "    end\n",
    "\n",
    "-----------------\n",
    "In BioRuby's command line it looks like this:\n",
    "\n",
    "\n",
    "<code>\n",
    "bioruby> Bio::FlatFile.open('./files/SP1.fq').each do |entry|\n",
    "**bioruby+**   hq = entry.mask(20)               < NOTE:  'bioruby+' means that it is waiting for more input\n",
    "puts hq.output_fasta(entry.entry_id)\n",
    "end\n",
    "<code>    \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Command line as a debugger\n",
    "\n",
    "One possible use of the BioRuby command line is as a debugger - you can explore an object \"live\" until you decide what command you want to use, then you put that command into your code.\n",
    "\n",
    "\n",
    "# END OF  COMMAND LINE\n",
    "\n",
    "For the rest of the course, we will use BioRuby inside of Jupyter just as we have before"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "----------------------\n",
    "\n",
    "----------------------\n",
    "\n",
    "-----------------------\n",
    "\n",
    "\n",
    "\n",
    "# BioRuby - Documentation\n",
    "\n",
    "http://bioruby.org/rdoc/\n",
    "\n",
    "# Sequences and Sequence Features\n",
    "\n",
    "We will focus on Sequences in this class.  \n",
    "* how to retrieve them\n",
    "* how to write them to files\n",
    "* how to change their formats\n",
    "* how to explore their features (e.g. genes)\n",
    "* how to create databases from them to:\n",
    "* make them easier to find/query\n",
    "\n",
    "# BioRuby's Sequence objects\n",
    "\n",
    "Like all other Ruby objects, you create a new Sequence object using \"new\"\n",
    "\n",
    "First, look at the documentation for Bio::Sequence, Bio::Sequence::AA, Bio::Sequence::NA (http://bioruby.org/rdoc/Bio/Sequence.html and find the other two yourself :-) )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "require 'bio'\n",
    "\n",
    "puts \"MYSEQ0 = Bio::Sequence.new('ACTTTGC')\"\n",
    "myseq0 = Bio::Sequence.new(\"ACTTTGC\")\n",
    "puts myseq0.class\n",
    "puts myseq0.seq\n",
    "puts myseq0.seq.class\n",
    "puts\n",
    "\n",
    "puts \"MYSEQ1 = Bio::Sequence.auto('ACTTTGC')\"\n",
    "myseq1 = Bio::Sequence.auto(\"ACTTTGC\")\n",
    "puts myseq1.class\n",
    "puts myseq1.seq\n",
    "puts myseq1.seq.class\n",
    "puts\n",
    "\n",
    "puts \"MYSEQ2 = Bio::Sequence::NA.new('ACTTTGC')\"\n",
    "myseq2 = Bio::Sequence::NA.new(\"ACTTTGC\")\n",
    "puts myseq2.class\n",
    "puts myseq2.seq\n",
    "puts myseq2.seq.class\n",
    "puts\n",
    "\n",
    "#puts myseq2.public_methods.join(\"\\n\")\n",
    "puts \"myseq2 = #{myseq2}\"\n",
    "\n",
    "myseq3 = myseq2.reverse_complement\n",
    "puts \"myseq3 = #{myseq3}  (reverse complement)\"\n",
    "\n",
    "myseq2.reverse_complement!    #   <---- note the ! ->  When you see this in Ruby, it means the method will change the object\n",
    "puts \"myseq2 = #{myseq2} (reverse complement!)\"\n",
    "\n",
    "# this is the better way to do it:\n",
    "puts myseq2.to_s    # to_s means \"to string\" - you will see this in a lot of Ruby objects\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Limitations\n",
    "\n",
    "* BioRuby does not have built-in connections to every database on earth (e.g. no connection to TAIR or AraPort)\n",
    " * You will still need to use REST interfaces and regular expressions...often!\n",
    "* BioRuby does not recognize every kind of sequence identifier (e.g. At1g287748 means nothing to BioRuby)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Object Methods\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "require 'bio'\n",
    "\n",
    "aaseq = Bio::Sequence.auto(\"BBSTD\")\n",
    "puts aaseq.seq.class\n",
    "puts aaseq.seq.public_methods.sort\n",
    "puts\n",
    "aaseq = Bio::Sequence.auto(\"AAAACCCCCCCTTTTTTTGGGGGG\")\n",
    "puts aaseq.seq.class\n",
    "puts aaseq.seq.public_methods.sort\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## writing sequences\n",
    "\n",
    "* :output \n",
    "* :list_output_formats\n",
    "* :output_fasta\n",
    "\n",
    "Note that the documentation for Bio::Sequence says:\n",
    "\n",
    "    Included Modules\n",
    "\n",
    "        Bio::Sequence::Format\n",
    "        Bio::Sequence::SequenceMasker \n",
    "\n",
    "That's where these \"output\" methods are coming from...\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aaseq.list_output_formats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "puts aaseq.output_fasta    # ask for fasta\n",
    "puts aaseq.output    # the default is fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "puts aaseq.output(:fastq)   # specify that you want fastq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## searching sequences \n",
    "\n",
    "see:  https://www.bioinformatics.org/sms/iupac.html\n",
    "\n",
    "This is accomplished in bioruby using the .to_re\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "require 'bio'\n",
    "\n",
    "seq = Bio::Sequence::NA.new(\"actgggggatccc\")\n",
    "search = Bio::Sequence::NA.new(\"ATMM\")   # M indicates it is an A or C\n",
    "\n",
    "re = Regexp.new(search.to_re)\n",
    "puts re.source   # to see the content of a regualr expression call .source\n",
    "\n",
    "match = seq.seq.match(re)\n",
    "puts match\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Other common sequence methods\n",
    "(from the BioRuby Tutorial)\n",
    "\n",
    "    seq = Bio::Sequence::NA.new(\"atgcatgcaaaa\")\n",
    "    puts seq.complement\n",
    "\n",
    "    puts seq.subseq(3,8) # gets subsequence of positions 3 to 8 (starting from 1)\n",
    "    puts seq.gc_percent \n",
    "    puts seq.composition \n",
    "    puts seq.translate \n",
    "    puts seq.translate(2)        # translate from frame 2\n",
    "    puts seq.translate(1,11)     # codon table 11\n",
    "    puts seq.translate.codes\n",
    "    puts seq.translate.names\n",
    "    puts seq.translate.composition\n",
    "    puts seq.translate.molecular_weight\n",
    "    puts seq.complement.translate\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = Bio::Sequence::NA.new(\"atgcatgcaaaa\")\n",
    "puts seq.complement\n",
    "\n",
    "puts seq.subseq(3,8) # gets subsequence of positions 3 to 8 (starting from 1)\n",
    "puts seq.gc_percent \n",
    "puts seq.composition \n",
    "puts seq.translate \n",
    "puts seq.translate(2)        # translate from frame 2\n",
    "puts seq.translate(1,11)     # codon table 11\n",
    "puts seq.translate.codes\n",
    "puts seq.translate.names\n",
    "puts seq.translate.composition\n",
    "puts seq.translate.molecular_weight\n",
    "puts seq.complement.translate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading common database formats\n",
    "\n",
    "BioRuby can import a wide range of Sequence files, in many formats:\n",
    "\n",
    "    $ ls -l /home/osboxes/.rvm/gems/ruby-2.4.2/gems/bio-1.5.1/lib/bio/db\n",
    "    total 324\n",
    "    -rw-r--r-- 1 osboxes osboxes  7198 Oct 26 05:58 aaindex.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 biosql\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 embl\n",
    "    -rw-r--r-- 1 osboxes osboxes 15444 Oct 26 05:58 fantom.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 fasta\n",
    "    -rw-r--r-- 1 osboxes osboxes  9384 Oct 26 05:58 fasta.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 fastq\n",
    "    -rw-r--r-- 1 osboxes osboxes 18543 Oct 26 05:58 fastq.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 genbank\n",
    "    -rw-r--r-- 1 osboxes osboxes 61643 Oct 26 05:58 gff.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes 10036 Oct 26 05:58 go.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 kegg\n",
    "    -rw-r--r-- 1 osboxes osboxes  9021 Oct 26 05:58 lasergene.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes  1478 Oct 26 05:58 litdb.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes  7622 Oct 26 05:58 medline.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes  5328 Oct 26 05:58 nbrf.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes 12178 Oct 26 05:58 newick.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes 56187 Oct 26 05:58 nexus.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 pdb\n",
    "    -rw-r--r-- 1 osboxes osboxes   517 Oct 26 05:58 pdb.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 phyloxml\n",
    "    -rw-r--r-- 1 osboxes osboxes 11233 Oct 26 05:58 prosite.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes 14132 Oct 26 05:58 rebase.rb\n",
    "    drwxrwxr-x 2 osboxes osboxes  4096 Oct 26 05:58 sanger_chromatogram\n",
    "    -rw-r--r-- 1 osboxes osboxes 14249 Oct 26 05:58 soft.rb\n",
    "    -rw-r--r-- 1 osboxes osboxes  6022 Oct 26 05:58 transfac.rb\n",
    "    \n",
    "(NOTE: some of these are folders, if the database contains different things - e.g. kegg has genes, pathways, compounds, etc.)\n",
    "\n",
    "BioRuby is also able to guess (usually correctly!) what kind of sequence file you give it.  \n",
    "\n",
    "So, you can either tell it what kind of data you have:\n",
    "\n",
    "---------------\n",
    "<code>\n",
    "    ff = Bio::FlatFile.new(Bio::GenBank, data)\n",
    "</code>\n",
    "----------------\n",
    "\n",
    "Or, you can ask BioRuby to look at the data and make the decision itself:\n",
    "\n",
    "----------------------\n",
    "<code>\n",
    "    ff = Bio::FlatFile.auto(data)\n",
    "\n",
    "    ff.each_entry do |entry|    # many database formats, like FASTA, allow multiple records in a single file/string\n",
    "      p entry.entry_id          # identifier of the entry\n",
    "      p entry.definition        # definition of the entry\n",
    "      p entry.seq               # sequence data of the entry\n",
    "    end\n",
    "</code>\n",
    "----------------------\n",
    "\n",
    "## Let's try this for ourselves:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "require 'net/http'   # this is how you access the Web\n",
    "require 'bio'\n",
    "address = URI('http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=At3g54340')  # create a \"URI\" object (Uniform Resource Identifier: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier)\n",
    "\n",
    "response = Net::HTTP.get_response(address)  # use the Net::HTTP object \"get_response\" method\n",
    "                                             \n",
    "record = response.body\n",
    "puts record[1..100]  # show just a little bit...\n",
    "\n",
    "# create a local file with this data\n",
    "File.open('At3g54340.embl', 'w') do |myfile|  # w makes it writable\n",
    "  myfile.puts record\n",
    "end\n",
    "\n",
    "\n",
    "puts \"Trying Method 1a - directly create a Bio::EMBL object from string\"\n",
    "#  METHOD 1a - Create an object of the correct type from a string\n",
    "entry = Bio::EMBL.new(record)\n",
    "puts entry.class\n",
    "puts \"The record is #{entry.definition} \"\n",
    "\n",
    "\n",
    "\n",
    "#puts \"\\n\\nTrying Method 1b - directly create a Bio::EMBL object from a file\"\n",
    "# METHOD 1b - Create an object of the correct type from a file\n",
    "#entry = Bio::EMBL.open('At3g54340.embl')\n",
    "#puts entry.class\n",
    "#puts \"The record is #{entry.definition} \"\n",
    "\n",
    "\n",
    "# ==========================================================\n",
    "# see documentation at http://bioruby.org/rdoc/Bio/FlatFile\n",
    "\n",
    "puts \"\\n\\nTrying method 2 - create a new FlatFile object of the correct type\"\n",
    "\n",
    "# we can ask what type it is...\n",
    "datafiletype = Bio::FlatFile.autodetect(record)\n",
    "puts datafiletype\n",
    "\n",
    "datafile1 = Bio::FlatFile.new(datafiletype, File.open('At3g54340.embl', 'r')) # use that to create the correct type\n",
    "# note that the .new requires a \"stream\" as its second argument.  This is ugly, but not difficult...\n",
    "\n",
    "puts datafile1.class \n",
    "\n",
    "datafile1.each_entry do |entry|   # the FILE is not the same as the RECORD - multiple records can exist in a file\n",
    "  puts entry.class\n",
    "  puts \"The record is #{entry.definition} \"\n",
    "end\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# ==========================================================\n",
    "# see documentation at http://bioruby.org/rdoc/Bio/FlatFile\n",
    "\n",
    "puts \"\\n\\nTrying method 3 - create a new FlatFile object using auto-detect for type\"\n",
    "datafile2 = Bio::FlatFile.auto('At3g54340.embl')\n",
    "puts datafile2.class  \n",
    "\n",
    "datafile2.each_entry do |entry| # the FILE is not the same as the RECORD - multiple records can exist in a file\n",
    "  puts entry.class\n",
    "  puts \"The record is *#{entry.definition}* \" unless entry.definition.nil? || entry.definition.empty?  # see this? Useful!\n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Methods available for the common Sequence files:\n",
    "\n",
    "http://bioruby.org/rdoc/Bio/EMBL\n",
    "\n",
    "\n",
    "    #cc\n",
    "    #comment\n",
    "    #data_class\n",
    "    #date_created\n",
    "    #date_modified\n",
    "    #dblinks\n",
    "    #division\n",
    "    #dt\n",
    "    #each_cds\n",
    "    #each_gene\n",
    "    #entry  *******\n",
    "    #entry_id\n",
    "    #entry_name\n",
    "    #entry_version\n",
    "    #features    ***********\n",
    "    #fh\n",
    "    #ft\n",
    "    #id_line\n",
    "    #molecule\n",
    "    #molecule_type\n",
    "    #naseq\n",
    "    #ntseq\n",
    "    #os\n",
    "    #release_created\n",
    "    #release_modified\n",
    "    #seq\n",
    "    #seqlen\n",
    "    #sequence_length\n",
    "    #species    ************\n",
    "    #sq\n",
    "    #sv\n",
    "    #to_biosequence\n",
    "    #topology\n",
    "    #version \n",
    "    \n",
    "    http://bioruby.org/rdoc/Bio/EMBLDB/Common.html\n",
    "    \n",
    "    #ac\n",
    "    #accession   *********\n",
    "    #accessions\n",
    "    #de\n",
    "    #definition\n",
    "    #description\n",
    "    #dr\n",
    "    #keywords\n",
    "    #kw\n",
    "    #oc\n",
    "    #og\n",
    "    #os\n",
    "    #ref\n",
    "    #references \n",
    "    \n",
    "    http://bioruby.org/rdoc/Bio/DB.html\n",
    "    \n",
    "    #entry_id\n",
    "    #exists?\n",
    "    #fetch\n",
    "    #get\n",
    "    #tags "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence features - exons, introns, polyA sites, etc.\n",
    "\n",
    "http://bioruby.org/rdoc/Bio/Feature.html\n",
    "\n",
    "http://bioruby.org/rdoc/Bio/Feature/Qualifier.html\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "require 'bio' \n",
    "\n",
    "datafile2 = Bio::FlatFile.auto('At3g54340.embl')\n",
    "puts datafile2.class  \n",
    "\n",
    "datafile2.each_entry do |entry| # the FILE is not the same as the RECORD - multiple records can exist in a file\n",
    "# shows accession and organism\n",
    "  next unless entry.accession\n",
    "\n",
    "  puts entry.class\n",
    "  puts \"# #{entry.accession} - #{entry.species}\"\n",
    "\n",
    "  # iterates over each element in 'features'\n",
    "  entry.features.each do |feature|\n",
    "    position = feature.position\n",
    "    puts \"\\n\\n\\n\\nPOSITION = #{position}\"\n",
    "    qual = feature.assoc            # feature.assoc gives you a hash of Bio::Feature::Qualifier objects \n",
    "                                    # i.e. qualifier['key'] = value  for example qualifier['gene'] = \"CYP450\")\n",
    "    puts \"Associations = #{qual}\"\n",
    "    # skips the entry if \"/translation=\" is not found\n",
    "    next unless qual['translation']    # this is an indication that the feature is a transcript\n",
    "\n",
    "    # collects gene name and so on and joins it into a string\n",
    "    gene_info = [\n",
    "      qual['gene'], qual['product'], qual['note'], qual['function']\n",
    "    ].compact.join(', ')\n",
    "    puts \"TRANSCRIPT FOUND!\\nGene Info:  #{gene_info}\"\n",
    "    # shows nucleic acid sequence\n",
    "    puts \"\\n\\n>NA splicing('#{position}') : #{gene_info}\"\n",
    "    puts entry.naseq.class   # this is a Bio::Sequence::NA    Look at the documentation to understand the .splicing() method\n",
    "    puts entry.naseq.splice(position)  # http://bioruby.org/rdoc/Bio/Sequence/Common.html#method-i-splice\n",
    "\n",
    "    # shows amino acid sequence translated from nucleic acid sequence\n",
    "    puts \"\\n\\n>AA translated by splicing('#{position}').translate\"\n",
    "    puts entry.naseq.splicing(position).translate\n",
    "\n",
    "    # shows amino acid sequence in the database entry (/translation=)\n",
    "    puts \"\\n\\n>AA original translation\"\n",
    "    puts qual['translation']\n",
    "  end\n",
    "\n",
    "  \n",
    "  puts \"\\n\\nNumber of features #{entry.features.length}\"\n",
    "  \n",
    "\n",
    "end\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "# Creating your own features\n",
    "\n",
    "The BioRuby Class that allows you to add new features is the Bio::Sequence class.  Other classes, like Bio::EMBL, are representations of DATABASE RECORDS - you don't manipulate EMBL records!  But you can \"clone\" these into a Bio::Sequence object, which allows you to add new features.\n",
    "\n",
    "See the code below to understand this more deeply:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "require 'bio' \n",
    "\n",
    "datafile2 = Bio::FlatFile.auto('At3g54340.embl')\n",
    "puts \"datafile 2 is of Class: #{datafile2.class}\\n\\n\\n\"\n",
    "\n",
    "\n",
    "entry =  datafile2.next_entry   # this is a way to get just one entry from the FlatFile\n",
    "\n",
    "puts \"working on #{entry.accession}\"\n",
    "puts \"entry is of class #{entry.class}\"\n",
    "puts \"This entry has:#{entry.features.length} features right now\"\n",
    "puts \"entry can respond to a call to RETRIEVE features? #{entry.respond_to?('features')}\"\n",
    "puts puts \"entry can respond to a call to SET features? #{entry.respond_to?('features=')} <-------!!!\"\n",
    "\n",
    "\n",
    "puts \"\\n\\nconverting it to a Bio::Sequence\"\n",
    "bioseq = entry.to_biosequence  # this is how you convert a database entry to a Bio::Sequence\n",
    "\n",
    "puts \"bioseq is of class #{bioseq.class}\"\n",
    "puts \"the equivalent of Bio::EMBL.accession is Bio::Sequence.primary_accession:  #{bioseq.primary_accession}\\n\\t\\t\\t\\t (I agree, that is a bit frustrating ;-) )\"\n",
    "puts \"This Bio::Sequence has:#{bioseq.features.length} features, the same as the Bio::EMBL - it is a clone, but....\"\n",
    "puts \"bioseq can respond to a call to RETRIEVE features? #{bioseq.respond_to?('features')}\"\n",
    "puts puts \"bioseq can respond to a call to SET features? #{bioseq.respond_to?('features=')} <-------!!!\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "  \n",
    "\n",
    "</pre>\n",
    "\n",
    "# Creating your own Bio::Features\n",
    "  \n",
    "You need to create a new [Bio::Feature](http://bioruby.org/rdoc/Bio/Feature.html) object:\n",
    "\n",
    "------------\n",
    "\n",
    "\n",
    "**Public Class Methods**\n",
    "\n",
    "*new(feature = '', position = '', qualifiers = [])*\n",
    "\n",
    "    Create a new Bio::Feature object. \n",
    "    \n",
    "Arguments:\n",
    "\n",
    "        (required) feature: type of feature (e.g. “exon”)\n",
    "\n",
    "        (required) position: position of feature (e.g. “complement(1532..1799)”)\n",
    "\n",
    "        (opt) qualifiers: list of Bio::Feature::Qualifier objects (default: [])\n",
    "\n",
    "Returns\n",
    "\n",
    "        Bio::Feature object\n",
    "\n",
    "\n",
    "**Public Instance Methods**\n",
    "\n",
    "*append(a)*\n",
    "\n",
    "    Appends a Qualifier object to the Feature.\n",
    "\n",
    "Arguments:\n",
    "\n",
    "        (required) qualifier: Bio::Feature::Qualifier object\n",
    "\n",
    "Returns\n",
    "\n",
    "        Bio::Feature object\n",
    "\n",
    "---------------\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "**NOTE:  You should read about [INSDC Feature tables](http://www.insdc.org/files/feature_table.html), <span style=\"color: red;\">especially how to format the location portion of a feature!!  [Location Examples](http://www.insdc.org/files/feature_table.html#3.4.3)**</span>\n",
    "\n",
    "</pre>\n",
    "\n",
    "</pre>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "require 'bio'  \n",
    "\n",
    "datafile2 = Bio::FlatFile.auto('At3g54340.embl')\n",
    "entry =  datafile2.next_entry   # this is a way to get just one entry from the FlatFile\n",
    "puts \"\\n\\nconverting it to a Bio::Sequence\"\n",
    "bioseq = entry.to_biosequence  # this is how you convert a database entry to a Bio::Sequence\n",
    "puts \"This entry has: #{bioseq.features.length} features at the beginning\"\n",
    "\n",
    "\n",
    "f1 = Bio::Feature.new('myrepeat','120..124')\n",
    "f1.append(Bio::Feature::Qualifier.new('repeat_motif', 'AAGCC'))\n",
    "f1.append(Bio::Feature::Qualifier.new('note', 'found by repeatfinder 2.0'))\n",
    "f1.append(Bio::Feature::Qualifier.new('strand', '+'))\n",
    "bioseq.features << f1  # you can append features one-by-one, using the << operator of Ruby arrays \n",
    "\n",
    "f2 = Bio::Feature.new('myrepeat','complement(190..194)')   # NOTE THE FORMAT HERE!  See note in RED above!!!!!!!!!!\n",
    "f2.append(Bio::Feature::Qualifier.new('repeat_motif', 'AAGCC'))\n",
    "f2.append(Bio::Feature::Qualifier.new('note', 'found by repeatfinder 2.0'))\n",
    "f2.append(Bio::Feature::Qualifier.new('strand', '-'))\n",
    "bioseq.features << f2\n",
    "\n",
    "\n",
    "puts \"This entry has: #{bioseq.features.length} features afer appending two individual features\"\n",
    "\n",
    "bioseq.features.concat([ f1, f2 ])   # or you can take an array of features and concatenate with the .features array\n",
    "\n",
    "puts \"This entry has: #{bioseq.features.length} features after concatenating a list of two new features\"\n",
    "\n",
    "  bioseq.features.each do |feature|\n",
    "    featuretype = feature.feature\n",
    "    next unless featuretype == \"myrepeat\"\n",
    "    position = feature.position\n",
    "    puts \"\\n\\n\\n\\nFEATURE #{featuretype} @ POSITION = #{position}\"\n",
    "    qual = feature.assoc            # feature.assoc gives you a hash of Bio::Feature::Qualifier objects \n",
    "                                    # i.e. qualifier['key'] = value  for example qualifier['gene'] = \"CYP450\")\n",
    "    puts \"Associations = #{qual}\"\n",
    "    # skips the entry if \"/translation=\" is not found\n",
    "  end\n",
    "\n",
    "puts \"\\n\\n\\ndone\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prove that you understand\n",
    "\n",
    "easy:  Modify the code above to print a list of the SNPs in this EMBL record.\n",
    "\n",
    "harder:  For each SNP, report if it is in an exon or an intron\n",
    "\n",
    "NOTE: from the Bio::Feature documentation\n",
    "\n",
    "\n",
    "### Attributes\n",
    "    feature[RW]\n",
    "    Returns type of feature in String (e.g 'CDS', 'gene')\n",
    "    \n",
    "    position[RW]\n",
    "    Returns position of the feature in String (e.g. 'complement(123..146)')\n",
    "\n",
    "    qualifiers[RW]\n",
    "    Returns an Array of Qualifier objects.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# General Feature Format - a concise way to represent sequence features\n",
    "\n",
    "It is common to want to exchange/publish sequence annotations.  There are many ways to do this - EMBL files, GenBank files, etc.; however, these formats are extremely \"rich\" and complicated.  The bioinformatics community has, for many years, looked for a \"quick and dirty\" way to exchange sequence features.  The solution they came to was called \"General Feature Format\" or \"GFF\".  GFF, basically, includes basic information about a sequence feature on a single line, consisting of 9 tab-separated fields:  \n",
    "    \n",
    "    seqid   source   type   start    end    score    strand    phase    attributes\n",
    "    \n",
    "These fields allow you to say what kind of feature, appears in what position, on which strand of which sequence, along with some additional metadata and annotations.\n",
    "\n",
    "Unfortunately, \"quick and dirty\" was not a very good idea!  For example, there was no agreement in the community about what types of features there could be, or how you could group features together (e.g. the exons of a specific gene).  Basically, it ended up not being very useful as a way to **exchange** feature data, because there was no agreement about... anything! So GFF went through a few iterations to try to solve these problems.\n",
    "\n",
    "There have been 3 increasingly rigorous versions of GFF created over the past ~10 years:  GFF, GFF2 and GFF3.  We are going to look at GFF3 - the most complex - because it fits with the discussion of ontologies we had a few weeks ago.\n",
    "\n",
    "The definition of the GFF3 Format is HERE: http://www.ensembl.org/info/website/upload/gff3.html\n",
    "\n",
    "The abbreviated explanation is:\n",
    "\n",
    "-----------\n",
    "\n",
    "\n",
    "\n",
    "**Fields**\n",
    "\n",
    "    The first line of a GFF3 file must be a comment that identifies the version, e.g.\n",
    "\n",
    "        ##gff-version 3\n",
    "\n",
    "Fields must be tab-separated. Also, all but the final field in each feature line must contain a value; \"empty\" columns should be denoted with a '.'\n",
    "\n",
    "  - **seqid** - name of the chromosome or scaffold; chromosome names can be given with or without the 'chr' prefix. Important note: the seq ID must be one used within Ensembl, i.e. a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.\n",
    "  - **source** - name of the program that generated this feature, or the data source (database or project name)\n",
    "  -  **type** - type of feature. Must be a term or accession from the **[SOFA sequence ontology](https://github.com/The-Sequence-Ontology/SO-Ontologies/blob/master/so.obo)**\n",
    "  -  **start** - Start position of the feature, with sequence numbering starting at 1.\n",
    "  -  **end** - End position of the feature, with sequence numbering starting at 1.\n",
    "  -  **score** - A floating point value.\n",
    "  -  **strand** - defined as + (forward) or - (reverse).\n",
    "  -  **phase** - One of '0', '1' or '2'. '0' indicates that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on..\n",
    "  - **attributes** - A semicolon-separated list of tag-value pairs, providing additional information about each feature. Some of these tags are predefined, e.g. ID, Name, Alias, Parent - see the GFF documentation for more details.\n",
    "\n",
    "The part I want you to pay attention to is \"**type**\" above.  It says that the \"type\" field is constrained to be a valid term from the Sequence Ontology.  You can [browse the Sequence Ontology here](http://www.sequenceontology.org/browser/obob.cgi).   Note that this is a **simple** explanation of the GFF3 format.  [The complete explanation is much more detailed](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md)!!\n",
    "\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "A typical line of GFF3 looks like this:\n",
    "\n",
    "     Chr3    GeneScan     coding_exon     1300     9745     0.97     +     1     ID=mrna0001.1;Name=sonichedgehog.e1\n",
    "\n",
    "This says that, on the sequence called Chr3 there is a [coding_exon](http://www.sequenceontology.org/browser/current_svn/term/SO:0000195) that runs on the + strand from sequence position 1300 to 1745 that was discovered by the GeneScan software, that gave it a liklihood score of 97%, and is in Prame-1 relative to the coding sequence (i.e. starts with the second nucleotide of the triplet-codon).  It has an ID of mrna0001.1, and it has a display-name of \"sonichedgehog.e1\"\n",
    "\n",
    "This may look trivial - and in fact, it is quite a simple format! - however, it can be difficult to create these files.  For example, the 'phase' field can be tricky to calculate.  Most importantly, the attributes field is used to create parent-child relationships (for example, Exon with ID E1.1 belongs to Transcript T1.1 which belongs to Gene G1) and these are quite difficult to both create and parse.  A larger example of [a Rice (_Oryza sativa_) GFF3 file is here]( http://ftp.gramene.org/archives/PAST_RELEASES/release45/data/vcf/oryza_sativa/Duitama/WGSOryza_CIAT_LSU_USDA_NCGR_SV/MV89-80_bowtie2_NGSEP_SV.gff)\n",
    "\n",
    "For my exercises, I want you to not worry about the Attributes field, and you may use a \".\" character in the \"phase\" field (\".\" means \"I don't know/don't care\").  The field I want you to pay attention to is the type field - when you need to refer to a type of sequence feature, I want you to look-up what the correct Sequence Ontology Term is. \n",
    "\n",
    "Finally, at the beginning of the GFF3 file, you must have a metadata line \"##gff-version 3\".  At the end of the GFF3, following the directive \"##FASTA\" on a single line, you **may** (optional) include the FASTA-formatted sequences corresponding to the features in the GFF file.  The FASTA identifier (in the >header of the FASTA record) should be identical to the identifier in Field #1 of the GFF.  For example:\n",
    "\n",
    "\n",
    "    ##gff-version 3\n",
    "    seq1.1   CutScan  restriction_enzyme_cut_site    3    8    .    +    .   ID=EcoRI_1;Name=EcoRI_CutSite_1\n",
    "    ##FASTA\n",
    "    >seq1.1\n",
    "    ATAGAATTCTTGCGATGGCGAGTTTACCGGCGTAT\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "  \n",
    "\n",
    "</pre>\n",
    "  \n",
    "# GFF and BioRuby\n",
    "\n",
    "\n",
    "In BioRuby, we can **read** GFF3 (but we cannot write it... unfortunately...)\n",
    "\n",
    "The three objects you need to understand are:\n",
    "\n",
    "- [Bio::GFF](http://bioruby.org/rdoc/Bio/GFF.html) --> the parent object, responds to #records method call\n",
    "\n",
    "- [Bio::GFF::GFF3](http://bioruby.org/rdoc/Bio/GFF/GFF3) - this represents a GFF **file** (all of the features, plus the metadata, plus the FASTA)\n",
    "\n",
    "- [Bio::GFF::GFF3::Record](http://bioruby.org/rdoc/Bio/GFF/GFF3/Record) - this represents a single line of the 9-field GFF format - i.e. a single record in the file.\n",
    "\n",
    "- [Bio::GFF::GFF2::Record](http://bioruby.org/rdoc/Bio/GFF/GFF2/Record) --> called by the GFF3::Record object - this gives you most of the method calls a GFF Record.\n",
    "\n",
    "There are a few frustrating aspects to the BioRuby GFF3 objects.  \n",
    "\n",
    "1. They consume a String (not a file or filehandle or other streaming object)\n",
    "2. They consume the entire string - all or nothing - so they use a LOT of memory!\n",
    "\n",
    "Regardless, we will continue...\n",
    "\n",
    "This is the code for parsing GFF into BioRuby objects:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "require 'bio'\n",
    "\n",
    "gff = <<END\n",
    "seq1.1\\tCutScan\\trestriction_enzyme_cut_site\\t3\\t8\\t.\\t+\\t.\\tID=EcoRI_1;Name=EcoRI_CutSite_1\n",
    "seq1.1\\tCutScan\\trestriction_enzyme_cut_site\\t10\\t15\\t.\\t+\\t.\\tID=EcoRI_1;Name=EcoRI_CutSite_1\n",
    "##FASTA\n",
    ">seq1.1\n",
    "ATAGAATTCTTGCGATGGCGAGTTTACCGGCGTAT\n",
    "\n",
    "END\n",
    "\n",
    "record = Bio::GFF::GFF3.new(gff)\n",
    "\n",
    "record.records.each do |feature|  # look at the Bio::GFF::GFF2 documentation to see the attributes\n",
    "  puts feature.class   # I am surprised that this is not a Seq::Feature object...???  Anyway...\n",
    "  puts feature.id  # a shortcut to the ID attribute (a requirement of GFF3)\n",
    "  print \"start and nucleotide #{feature.start} until nucl. #{feature.end} \"\n",
    "  print \"of type #{feature.feature} from #{feature.source} \"\n",
    "  puts  \"with name #{feature.get_attribute('Name')}\"  # NOTE you can call individual attributes by their name!\n",
    "  puts feature.attributes_to_hash\n",
    "\n",
    "  puts \"\\n\\n\"\n",
    "end\n",
    "\n",
    "puts \"the original record:\"\n",
    "puts record.to_s\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 2.5.1",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "2.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
